README.md
# gbench

This crate provides the tools to benchmark code for further analyzation using Chrome tracing. The crate is purely rust and has no dependencies.

# Advantages

- The crate is based on macros which are empty if debug_assertions is off
- Writing benchmarking data will have almost no effect on the time (1mcs per trace on my machine)
- Simple macro-based syntax
- You have control over how the collected data will be saved

# Example

For more examples and a guide on how to use please visit the [documentation](https://docs.rs/gbench)

```rust
use gbench::{instantiate, scope, ChromeTracing};
use std::thread;

fn calculate(num: f32, n: u32) -> f32 {
    (0..n)
        .fold((num, 0.0), |(x, v), _| (x + v * 0.01, v - x * 0.001))
        .0
}

fn main() {
    instantiate!(ChromeTracing("target/bench"));

    scope!(program_scope | "Program scope");

    // Doing the work that needs benchmarking
    for _ in 0..5 {
        scope!(main | "Main scope");

        // Spawning a thread to do work
        let thread = thread::spawn(move || {
            // This benchmarks the scope that it is in
            scope!(child | "Child");

            calculate(1.0, 1_500_000)
        });

        // You can organize your subtasks in scopes to
        // benchmark them
        scope!(imp | "An important task");

        {
            scope!(i1 | "Important subtask");
            calculate(1.0, 300_000);
        }

        {
            scope!(i2 | "Less important subtask");
            calculate(1.0, 500_000);
        }

        // If the block of code that you need to benchmark
        // has ended you can drop the guard if the scope
        // has not ended
        drop(imp);

        // Marking the start of another task
        scope!(join | "Joining thread");
        thread.join().unwrap();

        // This line of code is unnecessary but I like
        // to keep it
        drop(join);
    }
}
```

The code above produces this trace

![image](assets/bench.png)

# Links

- [crates.io](https://crates.io/crates/gbench)
- [crate documentation](https://docs.rs/gbench)


RELEASES.md
# Release 1.0.0

- Move to a writer system
- Added ChromeTracing and CsvWriter

# Release 0.3.1

- Doc updates

# Release 0.3.0

- Added count! macro

# Release 0.2.2

- Doc updates

# Release 0.2.1

- Improved benchmarking accuracy

# Release 0.2.0

- Added end method on struct Instantiator


src/test.rs
use std::fs::read_to_string;
use std::time::Duration;

use crate::{begin_gbench, end_gbench, send, BenchMessage};

#[test]
fn simple_write() {
    begin_gbench("target/bench.json", Duration::from_millis(100));
    send(BenchMessage::Log {
        log: String::from("Log"),
        ts: 0,
    });

    end_gbench();

    let contents = read_to_string("target/bench.json").unwrap();

    assert_eq!(&contents, "{\"otherData\":{},\"traceEvents\":[{\"cat\":\"log\",\"name\":\"start\",\"ph\":\"I\",\"pid\":0,\"tid\":0,\"ts\":0},{\"cat\":\"log\",\"name\":\"Log\",\"ph\":\"I\",\"pid\":0,\"tid\":0,\"ts\":0}]}");
}


src/id.rs
use std::sync::{Arc, Mutex};

type StorageInner = Arc<Mutex<Vec<bool>>>;

pub struct IdStorage(StorageInner);

impl IdStorage {
    pub fn new() -> IdStorage {
        IdStorage(Arc::new(Mutex::new(Vec::new())))
    }

    pub fn gen(&mut self) -> IdBorrow {
        let mut lock = self.0.lock().unwrap();
        let mut k = 0;
        while k < lock.len() && lock[k] {
            k += 1;
        }

        if k == lock.len() {
            lock.push(true);
        } else {
            lock[k] = true;
        }

        IdBorrow {
            storage: self.0.clone(),
            id: k,
        }
    }
}

pub struct IdBorrow {
    storage: StorageInner,
    id: usize,
}

impl IdBorrow {
    pub fn id(&self) -> usize {
        self.id
    }
}

impl Drop for IdBorrow {
    fn drop(&mut self) {
        let mut lock = self.storage.lock().unwrap();
        lock[self.id] = false;
    }
}


src/bench.rs
use std::mem;
use std::time::Instant;

use crate::global::{begin, begin_time, end, gen_id, get_id, queue_mutex, BenchData};
use crate::writer::Writer;

fn ts_of(instant: Instant) -> f32 {
    (instant.duration_since(begin_time()).as_nanos() as f32) / 1000.0
}

fn enqueue(data: BenchData) {
    let mut lock = queue_mutex();
    lock.push(data);
}

pub fn _log(log: String) {
    let ts = ts_of(Instant::now());
    let tid = get_id();

    enqueue(BenchData::Log { log, ts, tid });
}

fn bench(name: String, ts: f32) {
    let dur = ts_of(Instant::now()) - ts;
    let tid = get_id();

    enqueue(BenchData::Bench { name, ts, dur, tid });
}

pub fn _count(name: String, data: Vec<(String, f32)>) {
    let ts = ts_of(Instant::now());
    let tid = get_id();

    enqueue(BenchData::Count {
        name,
        data,
        ts,
        tid,
    });
}

/// Starts a benchmarking scope on creation and ends it on drop
///
/// TimeScope saves the Instant it was created. When dropped it
/// saves the benchmarking results to the file.
///
/// Using [scope!] macro instead of this struct is recommened.
///
/// [scope!]: macro.scope.html
pub struct TimeScope {
    start: f32,
    name: String,
}

impl TimeScope {
    pub fn new(name: String) -> TimeScope {
        gen_id();
        TimeScope {
            start: ts_of(Instant::now()),
            name,
        }
    }
}

impl Drop for TimeScope {
    fn drop(&mut self) {
        bench(mem::replace(&mut self.name, String::new()), self.start);
    }
}

/// Instantiates global data on creation and deinstantiates it on drop
///
/// This struct instantiates global data upon creation
/// and deinstantiates it upon drop. It also is responsible
/// for calling the writers when the data is collected.
///
/// Using [instantiate!] macro instead of this struct is recommened.
///
/// [instantiate!]: macro.instantiate.html
pub struct Instantiator {
    alive: bool,
    writers: Vec<Box<dyn Writer + 'static>>,
}

impl Instantiator {
    /// Constructs the instantiator
    ///
    /// The writers will be called in [end] method.
    ///
    /// [end]: struct.Instantiator.html#method.end
    pub fn new(writers: Vec<Box<dyn Writer + 'static>>) -> Instantiator {
        begin();
        Instantiator {
            alive: true,
            writers,
        }
    }

    /// Deinstantiates global variables and calls the writers
    ///
    /// This method is used when Instantiator is never dropped.
    // This method is called on drop.
    pub fn end(&mut self) {
        if self.alive {
            self.alive = false;
            end(mem::replace(&mut self.writers, Vec::new()));
        }
    }
}

impl Drop for Instantiator {
    fn drop(&mut self) {
        self.end();
    }
}


src/global.rs
use std::mem::{self, MaybeUninit};
use std::sync::{Mutex, MutexGuard};
use std::time::Instant;

use crate::id::{IdBorrow, IdStorage};
use crate::writer::Writer;

// Global data

/// Enum that represents different benchmarking data
///
/// Data that is passed to the writers is in form of these enums.
///
/// # Fields
/// - **ts** -  timestamp
/// - **dur** - duration
/// - **tid** - thread id
#[derive(Debug, Clone)]
pub enum BenchData {
    /// Log contains logging data produced by the [log!](macro.log.html) macro
    Log { log: String, ts: f32, tid: usize },

    /// Bench contains benchmarking data produced by the [scope!](macro.scope.html) macro
    Bench {
        name: String,
        ts: f32,
        dur: f32,
        tid: usize,
    },

    /// Count contains counting data produced by the [count!](macro.count.html) macro
    Count {
        name: String,
        ts: f32,
        tid: usize,
        data: Vec<(String, f32)>,
    },
}

static mut GLOBAL_DATA: MaybeUninit<GlobalData> = MaybeUninit::uninit();
struct GlobalData {
    pub program_begin: Instant,
    pub queue_mutex: Mutex<Vec<BenchData>>,
    pub id_storage: IdStorage,
}

// method for event loop starting
pub fn begin() {
    // data initialization
    let program_begin = Instant::now();

    // file mutex
    let queue_mutex = Mutex::new(Vec::new());

    // id storage
    let id_storage = IdStorage::new();

    // writing data to global
    unsafe {
        GLOBAL_DATA = MaybeUninit::new(GlobalData {
            queue_mutex,
            program_begin,
            id_storage,
        });
    }
}

// time of program beginning
pub fn begin_time() -> Instant {
    unsafe { (&*GLOBAL_DATA.as_ptr()).program_begin }
}

// getting queue mutex
pub fn queue_mutex() -> MutexGuard<'static, Vec<BenchData>> {
    unsafe { (&*GLOBAL_DATA.as_ptr()).queue_mutex.lock().unwrap() }
}

// deinstantiate the gobal data
pub fn end(writers: Vec<Box<dyn Writer + 'static>>) {
    // get data to write
    let data = {
        let mut lock = queue_mutex();
        mem::replace(&mut *lock, Vec::new())
    };

    // write data to writers
    for writer in writers {
        writer.end(&data);
    }

    // getting global data for dropping
    unsafe {
        let _gd = mem::replace(&mut GLOBAL_DATA, MaybeUninit::uninit()).assume_init();
    }
}

thread_local! {
    static TID: IdBorrow = unsafe { (&mut *GLOBAL_DATA.as_mut_ptr()).id_storage.gen() };
}

pub fn gen_id() {
    TID.with(|_| {});
}

pub fn get_id() -> usize {
    TID.with(|tid| tid.id())
}


src/writer.rs
use std::collections::HashSet;
use std::fs::File;
use std::time::{SystemTime, UNIX_EPOCH};
use std::io::Write;

use crate::global::BenchData;

/// The trait that is implemented by all the writers
///
/// # Examples
///
/// QueueLogger will log out to console all the collected data
/// when program ends.
/// ```rust
/// struct QueueLogger;
/// 
/// impl Writer for QueueLogger {
///     fn end(&self, data: &Vec<BenchData>) {
///         for data in data.iter() {
///             println!("{:#?}", data);
///         }
///     }
/// }
/// ```
pub trait Writer {
    /// This method is called on all the collected data when
    /// program ends
    fn end(&self, data: &Vec<BenchData>);
}

/// Writer for google chrome tracing
///
/// First field is the name of the folder where the 
/// output files will be saved.
pub struct ChromeTracing(pub &'static str);

fn write_data(file: &mut File, data: &BenchData) {
    match data {
        BenchData::Log { log, ts, tid } => write!(
            file,
            "{{\"cat\":\"log\",\"name\":\"{}\",\"ph\":\"I\",\"pid\":0,\"tid\":{},\"ts\":{}}}",
            log, tid, ts
        )
        .unwrap(),
        BenchData::Bench { name, ts, dur, tid } => write!(
            file,
            "{{\"cat\":\"function\",\"dur\":{},\"name\":\"{}\",\"ph\":\"X\",\"pid\":0,\"tid\":{},\"ts\":{}}}", 
            dur, name, tid,  ts
        ).unwrap(),
        BenchData::Count {name, ts, tid, data} => {
            write!(
                file, 
                "{{\"cat\":\"count\",\"name\":\"{}\",\"ph\":\"C\",\"pid\":0,\"tid\":{},\"ts\":{}, \"args\":{{", 
                name, tid, ts
            ).unwrap();

            let mut dataiter = data.into_iter();

            if let Some((name, value)) = dataiter.next() {
                write!(
                    file,
                    "\"{}\":{}",
                    name, value
                ).unwrap();
            }

            for (name, value) in dataiter {
                write!(
                    file, 
                    ",\"{}\":{}",
                    name, value
                ).unwrap();
            }

            write!(
                file, 
                "}}}}"
            ).unwrap();
        }
    }
}

impl Writer for ChromeTracing {
    fn end(&self, data: &Vec<BenchData>) {
        // write data to file
        let folder = self.0;
        let mut file = File::create(format!(
            "{}/bench-{}.json",
            folder,
            SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_millis()
        ))
        .unwrap();

        write!(file, "{{\"otherData\":{{}},\"traceEvents\":[").unwrap();

        let mut data = data.iter();

        // body
        if let Some(data) = data.next() {
            write_data(&mut file, data);
        }

        for data in data {
            write!(file, ",").unwrap();
            write_data(&mut file, data);
        }

        // write footer
        write!(file, "]}}").unwrap();
    }
}

/// Writer for csv format
///
/// This writer will save the counter data in form 
/// of a csv table.
///
/// First field is the name of the folder where the 
/// output files will be saved.
pub struct CsvWriter(pub &'static str);

const DELIMITER: char = ';';

fn tstr(v: f32) -> String {
    v.to_string().replace(".", ",")
}

impl Writer for CsvWriter {
    fn end(&self, data: &Vec<BenchData>) {
        let folder= self.0;
        let mut file = File::create(format!(
            "{}/graph-{}.csv", 
            folder,
            SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_millis())
        ).unwrap();

        let (fields, csvdata) = {
            let mut fields = HashSet::new();
            let mut csvdata = Vec::new();

            for bdata in data {
                if let BenchData::Count {name, ts, data, tid: _} = bdata {
                    for (varname, value) in data {
                        let fieldname = format!("{} : {}", name, varname);
                        csvdata.push((fieldname.clone(), *value, *ts));
                        fields.insert(fieldname);
                    }
                }
            }

            let mut fields = fields.into_iter().collect::<Vec<_>>();
            fields.sort();

            let csvdata = csvdata.into_iter().map(|(fieldname, value, ts)| {
                (fields.binary_search(&fieldname).unwrap(), value, ts)
            }).collect::<Vec<_>>();

            (fields, csvdata)
        };

        let fieldcount = fields.len();
        let rows = csvdata.into_iter().scan(vec![None;fieldcount], |state, (idx, value, ts)| {
            state[idx] = Some(value);
            Some((ts, state.clone()))
        });

        write!(file, "ts").unwrap();

        for field in fields {
            write!(file, "{}{}", DELIMITER, field).unwrap();
        }

        write!(file, "\n").unwrap();
        
        for (ts, data) in rows {
            write!(file, "{}", tstr(ts)).unwrap();
            
            for datapart in data {
                write!(file, "{}", DELIMITER).unwrap();
                if let Some(data) = datapart {
                    write!(file, "{}", tstr(data)).unwrap();
                }
            }
            
            write!(file, "\n").unwrap();
        }
    }
}

src/lib.rs
//! This crate provides the tools to benchmark code for
//! further analyzation using different tools.
//!
//! # Writers System
//!
//! This crate uses writers system to manage collected data.
//! This means that all the macros that collect data during
//! program running push this data in form of [BenchData] enums
//! to a shared storage. At the end of the program all the data is given
//! to the instances of [Writer] that are given at the initialization.
//!
//! # Examples
//!
//! Examples of using gbench basic functionality
//! ```rust
//! use gbench::{instantiate, scope, ChromeTracing};
//! fn main() {
//!     // Istantiation of the global variables
//!     // It is needed at the top of every program that uses gbench
//!     // The folder that is specified is the folder where the data
//!     // will be saved
//!     instantiate!(ChromeTracing("target/bench"));
//!     {
//!         // This macro creates a variable that starts benchmarking
//!         // on creation and saves the result on drop.
//!         // The variable name is the first argument, the scope name
//!         // is the second.
//!         scope!(sc | "Scope");
//!         for _ in 0..1_000_000 {
//!             let _a = 1 + 1;
//!         }
//!     }
//! }
//! ```
//!
//! Example of a [log!] macro use
//! ```rust
//! use gbench::{instantiate, log, scope, ChromeTracing};
//!
//! fn main() {
//!     instantiate!(ChromeTracing("target/bench"));
//!     {
//!         scope!(sc | "Scope");
//!         for _ in 0..1_000 {
//!             let a = 1 + 1;
//!             // You can log to the file with a timestamp
//!             // using log! macro
//!             log!("A = {}", a);
//!         }
//!     }
//! }
//! ```
//!
//! Example of a [count!] macro and [CsvWriter] writer use
//! ```rust
//! use gbench::{count, instantiate, scope, ChromeTracing, CsvWriter};
//!
//! fn main() {
//!     // Additionally CsvWriter will save all the counter data in
//!     // a csv table
//!     instantiate!(ChromeTracing("target/bench"), CsvWriter("target/bench"));
//!     {
//!         scope!(sc | "Scope");
//!         for i in 0..1000 {
//!             let val = i * i;
//!             // This statement writes val to field "val" of counter "Actual value"
//!             // and writes i to field "i" of counter "I"
//!             count! {
//!                 "Actual value" => {
//!                     "val" => val
//!                 },
//!                 "I" => {
//!                     "i" => i
//!                 }
//!             }
//!         }
//!     }
//! }
//! ```
//!
//! Full example
//! ```rust
//! use gbench::{instantiate, scope, ChromeTracing};
//! use std::thread;
//! fn calculate(num: f32, n: u32) -> f32 {
//!     (0..n)
//!         .fold((num, 0.0), |(x, v), _| (x + v * 0.01, v - x * 0.001))
//!         .0
//! }
//!
//! fn main() {
//!     instantiate!(ChromeTracing("target/bench"));
//!
//!     scope!(program_scope | "Program scope");
//!     // Doing the work that needs benchmarking
//!     for _ in 0..5 {
//!         scope!(main | "Main scope");
//!         // Spawning a thread to do work
//!         let thread = thread::spawn(move || {
//!             // This benchmarks the scope that it is in
//!             scope!(child | "Child");
//!             calculate(1.0, 1_500_000)
//!         });
//!         // You can organize your subtasks in scopes to
//!         // benchmark them
//!         scope!(imp | "An important task");
//!         {
//!             scope!(i1 | "Important subtask");
//!             calculate(1.0, 300_000);
//!         }
//!         {
//!             scope!(i2 | "Less important subtask");
//!             calculate(1.0, 500_000);
//!         }
//!         // If the block of code that you need to benchmark
//!         // has ended you can drop the guard if the scope
//!         // has not ended
//!         drop(imp);
//!         // Marking the start of another task
//!         scope!(join | "Joining thread");
//!         thread.join().unwrap();
//!         // This line of code is unnecessary but I like
//!         // to keep it
//!         drop(join);
//!     }
//! }
//! ```
//!
//! [log!]: macro.log.html
//! [count!]: macro.count.html
//! [CsvWriter]: struct.CsvWriter.html
//! [BenchData]: enum.BenchData.html
//! [Writer]: trait.Writer.html

mod bench;
mod global;
mod id;
mod writer;

pub use bench::Instantiator;
pub use bench::TimeScope;

pub use global::BenchData;
pub use writer::ChromeTracing;
pub use writer::CsvWriter;
pub use writer::Writer;

#[doc(hidden)]
pub use bench::_log;

#[doc(hidden)]
pub use bench::_count;

/// Benchmarks a scope of code
///
/// # Implementation
///
/// The macro expands into a [TimeScope] declaration
///
/// ```rust
/// scope!(main)
/// // expands into this
/// let main = TimeScope::new(format!("main"));
/// ```
///
/// ```rust
/// scope!(main | "A {}", 0)
/// // expands into this
/// let main = TimeScope::new(format!("A {}", 0));
/// ```
///
/// [TimeScope]: struct.TimeScope.html
///
/// # Examples
///
/// ```rust
/// // You can organize your subtasks in scopes to
/// // benchmark them
/// scope!(imp | "An important task");
///
/// {
///     scope!(i1 | "Important subtask");
///     do_work();
/// }
///
/// {
///     scope!(i2 | "Less important subtask");
///     do_other_work();
/// }
///
/// // If the block of code that you need to benchmark
/// // has ended you can drop the guard if the scope
/// // has not ended
/// drop(imp);
///
/// // rest of the scope...
/// ```
#[cfg(debug_assertions)]
#[macro_export]
macro_rules! scope {
    ($name:ident) => {
        scope!($name | stringify!(name));
    };

    ($name:ident | $($arg:tt)*) => {
        let $name = {
            use gbench::TimeScope;
            TimeScope::new(format!($($arg)*))
        };
    };
}

#[cfg(not(debug_assertions))]
#[macro_export]
macro_rules! scope {
    ($name:ident) => {};

    ($name:ident|$($arg:tt)*) => {};
}

/// Instantiates the global variables for benchmark logging
///
/// This macro should be used at the top of any program using this crate.
///
/// # Implementation
///
/// This macro expands into a declaration of [Instantiator] which instantiates
/// global variables on creation and deinstantiates them on drop.
///
/// ```rust
/// instantiate!(ChromeTracing("target/bench"));
/// // expands into this
/// let __gbench_instantiator__ = Instantiator::new(vec![Box::new(ChromeTracing("target/bench"))]);
/// ```
///
/// If you need to deinstantiate global variables before variable goes out
/// of scope you can specify the variable name and then call [end] on it
/// when you need the deinstantiation.
///
/// ```rust
/// instantiate!(ginst | ChromeTracing("target/bench"));
/// // expands into this
/// let ginst = Instantiator::new(vec![Box::new(ChromeTracing("target/bench"))]);
/// ```
///
/// [Instantiator]: struct.Instantiator.html
/// [end]: struct.Instantiator.html#method.end
///
/// # Examples
/// ```rust
/// use gbench::{instantiate, scope, ChromeTracing};
///
/// fn main() {
///     instantiate!(ChromeTracing("target/bench"));
///     {
///         scope!(sc | "Scope");
///
///         for _ in 0..1_000_000 {
///             let _a = 1 + 1;
///         }
///     }
/// }
/// ```
/// or using [end]
/// ```rust
/// use gbench::{instantiate, scope, ChromeTracing};
///
/// fn main() {
///     instantiate!(ginst | ChromeTracing("target/bench"));
///     {
///         scope!(sc | "Scope");
///
///         for _ in 0..1_000_000 {
///             let _a = 1 + 1;
///         }
///     }
///     ginst.end();
/// }
/// ```
#[cfg(debug_assertions)]
#[macro_export]
macro_rules! instantiate {
    ($name: ident | $($writer: expr),*) => {
        let mut $name = {
            use gbench::Instantiator;

            let mut writers = Vec::new();

            $(
                writers.push(Box::new($writer) as Box<dyn gbench::Writer + 'static>);
            )*

            Instantiator::new(writers)
        };
    };

    ($($writer: expr),*) => {
        instantiate!(__global_instantiator__ | $($writer),*);
    };
}

#[cfg(not(debug_assertions))]
#[macro_export]
macro_rules! instantiate {
    ($folder: expr) => {};
}

/// Logs data to a benchmarking file
///
/// ```rust
/// let a = 0;
/// log!("A: {}", a);
/// ```
/// will queue this [BenchData]
/// ```
/// Log {
///     log: "A: 0",
///     ts: /* event's timestamp */,
///     tid: /* event's thread of execution */,
/// }
/// ```
///
/// [BenchData]: enum.BenchData.html
#[cfg(debug_assertions)]
#[macro_export]
macro_rules! log {
    ($($arg:tt)*) => {
        {
            use gbench::_log as log;
            let log_string = format!($($arg)*);
            log(log_string);
        }
    };
}

#[cfg(not(debug_assertions))]
#[macro_export]
macro_rules! log {
    ($($arg:tt)*) => {};
}

/// Creates a counting event
///
/// ```rust
/// let i = 10;
/// count!(
///     "a" => {
///         "a" => i,
///         "b" => i / 2
///     },
///     "b" => {
///         "c" => i % 3
///     }
/// );
/// ```
///
/// Will queue these [BenchData]
///
/// ```
/// Count {
///     name: "a",
///     ts: /* event's timestamp */,
///     tid: /* event's thread of execution */,
///     data: [
///         (
///             "a",
///             10.0,
///         ),
///         (
///             "b",
///             5.0,
///         ),
///     ],
/// },
/// Count {
///     name: "b",
///     ts: /* event's timestamp */,
///     tid: /* event's thread of execution */,
///     data: [
///         (
///             "c",
///             1.0,
///         ),
///     ],
/// },
/// ```
///
/// [BenchData]: enum.BenchData.html
#[cfg(debug_assertions)]
#[macro_export]
macro_rules! count {
    ($($name: expr => {$($argname: expr => $val:expr),*}),*) => {{
        use gbench::_count as count;

        $(
            let cname = String::from($name);

            let mut data = Vec::new();

            $(
                data.push((String::from($argname), $val as f32));
            )*

            count(cname, data);
        )*
    }};
}

#[cfg(not(debug_assertions))]
#[macro_export]
macro_rules! count {
    ($name: ident => {$($argname: ident => $val: expr),*}) => {};
}
